File: runner/runner.c
#include <stdio.h>
#include <stdlib.h>
#include <windows.h>
#include <tchar.h>
#include <winreg.h>

#include "../bib/defines.h"

// Run as admin: Start-Process "runner.exe" -Verb runAs

/* Function to get com0com installation directory from the Windows Registry */
char* get_com0com_install_directory() {
    HKEY hKey;
    static char installDir[MAX_PATH];
    DWORD bufLen = MAX_PATH;
    
    if (RegOpenKeyEx(HKEY_LOCAL_MACHINE, 
                     "SOFTWARE\\WOW6432Node\\com0com", 
                     0, 
                     KEY_QUERY_VALUE, 
                     &hKey) == ERROR_SUCCESS) {
        RegQueryValueEx(hKey, "Install_Dir", NULL, NULL, (LPBYTE)installDir, &bufLen);
        RegCloseKey(hKey);
        return installDir;
    } else {
        DEBUG_ERROR("Could not open registry key for com0com.\n");
        return NULL;
    }
}

/* Function to execute a command and capture the output */
int exec_command(char* cmd, char* output, int output_size) {
    if (cmd == NULL || output == NULL) {
        DEBUG_ERROR("NULL pointers! \n");
        return EXIT_FAILURE;
    }
    
    FILE* pipe = _popen(cmd, "r");
    if (!pipe) return -1;
    fgets(output, output_size, pipe);
    _pclose(pipe);

    return EXIT_SUCCESS;
}

/* Function to list all the open ports */
void list_ports(char* install_dir) {
    char command[512];
    int exit;

    DEBUG_INFO("Ports list:\n");
    snprintf(command, sizeof(command), "\"%s\\setupc.exe\" list", install_dir);

    exit = system(command);
    if (exit != 0) {
        DEBUG_ERROR("Error listing ports. Error code: %u", exit);
    }
}

// Function to check if ports COM8 and COM9 are open
int check_ports_open(char* result) {
    return (strstr(result, "COM8") != NULL && strstr(result, "COM9") != NULL);
}

/* Function to open ports COM8 and COM9 */
void open_ports(char* install_dir, char* port_a, char* port_b) {
    char command[512];
    SetCurrentDirectory(install_dir);

    DEBUG_INFO("Removing CNC3 ports.");
    snprintf(command, sizeof(command), "\"%s\\setupc.exe\" --silent remove 3", install_dir);
    if (system(command) == 0) {
        DEBUG_OK("CNC3 removed correctly.\n");
    } else {
        DEBUG_INFO("CNC3 not removed. Maybe not opened yet.\n");
    }
    
    snprintf(command, sizeof(command), "\"%s\\setupc.exe\" install 3 PortName=%s PortName=%s", install_dir, port_a, port_b);    
    if (system(command) == 0) {
        DEBUG_OK("Ports %s and %s have been successfully opened.\n", port_a, port_b);
        list_ports(install_dir);
    } else {
        DEBUG_ERROR("Error opening the ports.\n");
        return;
    }
}

int main(void) {
    int ret = EXIT_FAILURE;

    char* install_dir = get_com0com_install_directory();
    if (!install_dir) {
        DEBUG_ERROR("Failed to retrieve com0com installation directory.\n");
        goto end;
    }

    list_ports(install_dir);
    open_ports(install_dir, "COM8", "COM9");
    
    printf("\nPress any key to continue...\n");
    getchar();

    ret = EXIT_SUCCESS;
end:
    return ret;
}

File: runner/runner.py
import subprocess, os, sys, ctypes
from pathlib import Path
import re
import winreg

from defines import *

def is_admin():
    """Check if the script is running with administrator privileges."""
    try:
        return os.getuid() == 0  # Unix-based systems (Linux, macOS)
    except AttributeError:
        return ctypes.windll.shell32.IsUserAnAdmin() != 0  # Windows

def run_as_admin():
    """Prompt the user to rerun the script with administrator privileges if needed."""
    if not is_admin():
        print("This script requires administrator privileges. Please run it as an administrator.")
        if sys.platform == "win32":
            ctypes.windll.shell32.ShellExecuteW(None, "runas", sys.executable, " ".join(sys.argv), None, 1)
        else:
            print("Please rerun the script with 'sudo' on Linux or macOS.")
        sys.exit()

# Call the function at the start of the script
run_as_admin()

def get_com0com_install_directory():
    """Fetches and returns the installation directory of com0com."""
    with winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, 
                        R'SOFTWARE\WOW6432Node\com0com') as key:
        return Path(winreg.QueryValueEx(key, 'Install_Dir')[0])

com0com_dir = get_com0com_install_directory()

def get_ports(results):
    """Extracts the port numbers and their identifiers CNCAx and CNCBx from the given output.
    
    Args:
        results (str): The output from `setupc.exe list`.

    Returns:
        list: List of used port numbers.
    """
    # Search for all CNCAx and CNCBx port numbers
    ports = re.findall(r'CNCA(\d+)', results)
    return ports

def list_ports():
    """Lists all current ports."""
    command = [str(com0com_dir / 'setupc.exe'), 'list']
    try:
        result = subprocess.run(command, cwd=com0com_dir, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
        DEBUG_INFO("Listing all open ports:")
        print(result.stdout)
        if result.stderr:
            print("Errors:\n", result.stderr)
        return result.stdout
    except Exception as e:
        print(f"Error executing command: {e}")

def close_all_ports():
    """Closes all active com0com ports."""
    results = list_ports()
    
    if results is None:
        return
    
    # Extract the real port numbers
    port_numbers = get_ports(results)

    DEBUG_INFO("Ports detected for closing: %s", ', '.join(port_numbers))

    # Close all active ports
    for port_number in port_numbers:
        command_remove = [str(com0com_dir / 'setupc.exe'), 'remove', port_number]
        try:
            subprocess.run(command_remove, cwd=com0com_dir, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True, check=True)
            print(f"Port CNCA{port_number} and CNCB{port_number} successfully closed.")
        except subprocess.CalledProcessError as e:
            print(f"[-] Error closing ports CNCA{port_number} and CNCB{port_number}: {e.stdout} - {e.stderr}")

def open_ports(port_a='COM8', port_b='COM9'):
    """Opens two new virtual ports with the specified names."""
    command_install = [
        str(com0com_dir / 'setupc.exe'), 'install', 
        f'PortName={port_a}', f'PortName={port_b}'
    ]
    
    try:
        result = subprocess.run(command_install, cwd=com0com_dir, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
        if result.stderr:
            print("Errors:\n", result.stderr)
            return
        
        print(f"[+] Ports {port_a} and {port_b} have been successfully opened.")
    except Exception as e:
        print(f"[-] Error opening the ports: {e}")

if __name__ == "__main__":
    print("\n")
    results = list_ports()

    # Check if COM8 and COM9 are present in the list
    if 'COM8' in results and 'COM9' in results:
        DEBUG_INFO("Ports COM8 and COM9 are already open.")
        DEBUG_OK("Ports COM8 and COM9 are ready for use.")
    else:
        DEBUG_INFO("Ports COM8 and COM9 are not open. Opening now...")
        open_ports('COM8', 'COM9')
    print("\n")